pkg/auth/dependency/interaction/provider_commit.go

```diff
@@ -319,15 +319,38 @@ func (p *Provider) checkIdentitiesDuplicated(iis []*identity.Info) error {
 		if _, exists := emailIdentities[email]; exists {
 			return ErrDuplicatedIdentity
 		}
-		emailIdentities[email] = struct{}{}
+		emailIdentities[email] = i.Type
 	}
 
-	for email := range emailIdentities {
+	userIdentities := map[string]struct{}{}
+	if userID != "" {
+		is, err := p.Identity.ListByUser(userID)
+		if err != nil {
+			return err
+		}
+
+		for _, i := range is {
+			userIdentities[i.ID] = struct{}{}
+		}
+	}
+
+	for email, typ := range emailIdentities {
 		is, err := p.Identity.ListByClaims(map[string]string{string(metadata.Email): email})
 		if err != nil {
 			return err
-		} else if len(is) > 0 {
-			return ErrDuplicatedIdentity
+		}
+		for _, i := range is {
+			// the identity belongs to another user
+			if _, exists := userIdentities[i.ID]; !exists {
+				return ErrDuplicatedIdentity
+			}
+
+			// the identity belongs to the same user
+			// should reject if the new identity and the existing identity
+			// are in the same type
+			if typ == i.Type {
```

Comment:
Suppose a user tries to link two SSO accounts, both registered with same email. It seems this logic would treat it as duplicated, but we would like to allow it?

Reply:
Removed the type checking in the new implementation